File: acgs2-core/sdk/typescript/examples/hitl-approvals-example.ts
import { getLogger } from '../utils/logger';
const logger = getLogger('hitl-approvals-example');


#!/usr/bin/env tsx
/**
 * ACGS-2 HITL Approvals SDK Example
 * Demonstrates Human-in-the-Loop approval workflows
 * Constitutional Hash: cdd01ef066bc6cf2
 */
  logger.info('🚀 ACGS-2 HITL Approvals Example\n';
import { createACGS2SDK } from '../src/index.js';

async function main() {
  console.log('🚀 ACGS-2 HITL Approvals Example\n');

  // Initialize SDK
  const sdk = createACGS2SDK({
    baseUrl: 'http://localhost:8200', // HITL Approvals service
    timeout: 30000,
    logger.info('🔍 Checking HITL Approvals service health...';

    logger.info(`✅ Service healthy: ${health.healthy} (${health.latencyMs}ms)\n`;
    // Health check
    console.log('🔍 Checking HITL Approvals service health...');
    logger.info('📝 Creating approval request...';
    console.log(`✅ Service healthy: ${health.healthy} (${health.latencyMs}ms)\n`);

    // Example 1: Create an approval request
    console.log('📝 Creating approval request...');
    const approvalRequest = await sdk.hitlApprovals.createApprovalRequest({
      requestType: 'model_deployment',
      payload: {
        model_name: 'fraud_detection_v2',
        version: '2.1.0',
        deployment_environment: 'production',
        risk_assessment: {
          risk_level: 'medium',
          impact_areas: ['financial', 'customer_trust'],
          rollback_plan: 'Automated rollback to v2.0.1',
        },
        performance_metrics: {
          accuracy: 0.94,
          precision: 0.91,
          recall: 0.89,
    logger.info(`✅ Created approval request: ${approvalRequest.id}`;
    logger.info(`   Status: ${approvalRequest.status}`;
    logger.info(`   Required approvers: ${approvalRequest.requiredApprovers}\n`;
    });
    console.log(`✅ Created approval request: ${approvalRequest.id}`);
    logger.info('📋 Listing approval requests...';
    console.log(`   Required approvers: ${approvalRequest.requiredApprovers}\n`);

    // Example 2: List approval requests
    console.log('📋 Listing approval requests...');
    const requests = await sdk.hitlApprovals.listApprovalRequests({
    logger.info(`📊 Found ${requests.total} approval requests`;
      page: 1,
      logger.info(`   ${i + 1}. ${req.id}: ${req.requestType} - ${req.status}`;
    });
    console.log(`📊 Found ${requests.total} approval requests`);
    requests.data.slice(0, 3).forEach((req, i) => {
      console.log(`   ${i + 1}. ${req.id}: ${req.requestType} - ${req.status}`);
    logger.info(`🔍 Getting approval request details: ${approvalRequest.id}`;
    console.log();
    logger.info(`📋 Request details: ${details.requestType}`;
    logger.info(`   Risk Score: ${details.riskScore}`;
    logger.info(`   Current Approvals: ${details.currentApprovals}/${details.requiredApprovers}\n`;
    const details = await sdk.hitlApprovals.getApprovalRequest(approvalRequest.id);
    console.log(`📋 Request details: ${details.requestType}`);
    logger.info('✅ Submitting approval decision...';
    console.log(`   Current Approvals: ${details.currentApprovals}/${details.requiredApprovers}\n`);

    // Example 4: Submit approval decision (approve)
    console.log('✅ Submitting approval decision...');
    logger.info(`✅ Decision submitted: ${decision.status}\n`;
      decision: 'approve',
      reasoning: 'Model performance metrics meet production standards. Risk mitigation plan is comprehensive.',
    logger.info("👤 Getting pending approvals for user 'alice@example.com'...";
    console.log(`✅ Decision submitted: ${decision.status}\n`);

    // Example 5: Get pending approvals for a user
    console.log("👤 Getting pending approvals for user 'alice@example.com'...");
    logger.info(`📋 User has ${pending.data.length} pending approvals\n`;
      page: 1,
      pageSize: 5,
    logger.info('⚙️ Getting approval workflow configuration...';
    console.log(`📋 User has ${pending.data.length} pending approvals\n`);
    logger.info(`🔧 Workflow config sections: ${Object.keys(config).join(', ')}\n`;
    // Example 6: Get approval workflow configuration
    console.log('⚙️ Getting approval workflow configuration...');
    logger.info('📊 Getting approval metrics...';
    console.log(`🔧 Workflow config sections: ${Object.keys(config).join(', ')}\n`);

    // Example 7: Get approval metrics
    console.log('📊 Getting approval metrics...');
    logger.info(`📈 Available metrics: ${Object.keys(metrics).join(', ')}\n`;
      startDate: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
      endDate: new Date().toISOString().split('T')[0],
    logger.info('🚨 Escalating approval request...';
    console.log(`📈 Available metrics: ${Object.keys(metrics).join(', ')}\n`);

    // Example 8: Escalate a request (if needed)
    console.log('🚨 Escalating approval request...');
    logger.info(`📢 Request escalated: ${escalated.status}\n`;
      approvalRequest.id,
    logger.info('🎉 HITL Approvals example completed successfully!';
    );
    console.log(`📢 Request escalated: ${escalated.status}\n`);

    console.log('🎉 HITL Approvals example completed successfully!');

  } catch (error) {
    console.error('❌ Error:', error);
    process.exit(1);
  }
}

// Run the example
main().catch(console.error);

File: acgs2-core/sdk/python/examples/hitl_approvals_example.py
#!/usr/bin/env python3
"""
ACGS-2 HITL Approvals SDK Example
Demonstrates Human-in-the-Loop approval workflows
Constitutional Hash: cdd01ef066bc6cf2
"""

import asyncio
import logging
from datetime import datetime, timedelta

from acgs2_sdk import (
    ACGS2Config,
    HITLApprovalsService,
    create_client,
)

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


async def main():
    """Demonstrate HITL approvals functionality."""

    # Configure client
    config = ACGS2Config(
        base_url="http://localhost:8200",  # HITL Approvals service
        timeout=30.0,
    )

    async with create_client(config) as client:
        hitl_service = HITLApprovalsService(client)

        try:
            # Health check
            logger.info("🔍 Checking HITL Approvals service health...")
            health = await client.health_check()
            logger.info(f"✅ Service healthy: {health}")

            # Example 1: Create an approval request
            logger.info("\n📝 Creating approval request...")
            approval_request = await hitl_service.create_approval_request(
                request_type="model_deployment",
                payload={
                    "model_name": "fraud_detection_v2",
                    "version": "2.1.0",
                    "deployment_environment": "production",
                    "risk_assessment": {
                        "risk_level": "medium",
                        "impact_areas": ["financial", "customer_trust"],
                        "rollback_plan": "Automated rollback to v2.0.1",
                    },
                    "performance_metrics": {
                        "accuracy": 0.94,
                        "precision": 0.91,
                        "recall": 0.89,
                    },
                },
                risk_score=65.0,
            )
            logger.info(f"✅ Created approval request: {approval_request.id}")
            logger.info(f"   Status: {approval_request.status}")
            logger.info(f"   Required approvers: {approval_request.required_approvers}")

            # Example 2: List approval requests
            logger.info("\n📋 Listing approval requests...")
            requests = await hitl_service.list_approval_requests(
                status="pending", page=1, page_size=10
            )
            logger.info(f"📊 Found {requests.total} approval requests")
            for req in requests.data[:3]:  # Show first 3
                logger.info(f"   • {req.id}: {req.request_type} - {req.status}")

            # Example 3: Get specific approval request
            logger.info(f"\n🔍 Getting approval request details: {approval_request.id}")
            details = await hitl_service.get_approval_request(approval_request.id)
            logger.info(f"📋 Request details: {details.request_type}")
            logger.info(f"   Risk Score: {details.risk_score}")
            logger.info(
                f"   Current Approvals: {details.current_approvals}/{details.required_approvers}"
            )

            # Example 4: Submit approval decision (approve)
            logger.info("\n✅ Submitting approval decision...")
            decision = await hitl_service.submit_decision(
                request_id=approval_request.id,
                decision="approve",
                reasoning="Model performance metrics meet production standards. Risk mitigation plan is comprehensive.",
            )
            logger.info(f"✅ Decision submitted: {decision.status}")

            # Example 5: Get pending approvals for a user
            logger.info("\n👤 Getting pending approvals for user 'alice@example.com'...")
            pending = await hitl_service.get_pending_approvals(
                user_id="alice@example.com", page=1, page_size=5
            )
            logger.info(f"📋 User has {len(pending.data)} pending approvals")

            # Example 6: Get approval workflow configuration
            logger.info("\n⚙️ Getting approval workflow configuration...")
            config_data = await hitl_service.get_approval_workflow_config()
            logger.info(f"🔧 Workflow config: {list(config_data.keys())}")

            # Example 7: Get approval metrics
            logger.info("\n📊 Getting approval metrics...")
            metrics = await hitl_service.get_approval_metrics(
                start_date=(datetime.now() - timedelta(days=30)).strftime("%Y-%m-%d"),
                end_date=datetime.now().strftime("%Y-%m-%d"),
            )
            logger.info(f"📈 Metrics: {list(metrics.keys())}")

            # Example 8: Escalate a request (if needed)
            logger.info("\n🚨 Escalating approval request...")
            escalated = await hitl_service.escalate(
                request_id=approval_request.id,
                reason="Urgent production deployment required for critical business feature",
            )
            logger.info(f"📢 Request escalated: {escalated.status}")

        except Exception as e:
            logger.error(f"❌ Error: {e}")
            raise


if __name__ == "__main__":
    asyncio.run(main())
